enterprise/server/remote_execution/runner/runner_linux.go

```diff
@@ -0,0 +1,173 @@
+//go:build linux && !android
+
+package runner
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"os"
+	"strings"
+
+	"github.com/buildbuddy-io/buildbuddy/enterprise/server/remote_execution/container"
+	"github.com/buildbuddy-io/buildbuddy/enterprise/server/remote_execution/containers/bare"
+	"github.com/buildbuddy-io/buildbuddy/enterprise/server/remote_execution/containers/docker"
+	"github.com/buildbuddy-io/buildbuddy/enterprise/server/remote_execution/containers/firecracker"
+	"github.com/buildbuddy-io/buildbuddy/enterprise/server/remote_execution/containers/podman"
+	"github.com/buildbuddy-io/buildbuddy/enterprise/server/remote_execution/containers/sandbox"
+	"github.com/buildbuddy-io/buildbuddy/enterprise/server/remote_execution/platform"
+	"github.com/buildbuddy-io/buildbuddy/enterprise/server/util/vfs_server"
+	"github.com/buildbuddy-io/buildbuddy/server/environment"
+	"github.com/buildbuddy-io/buildbuddy/server/util/status"
+
+	fcpb "github.com/buildbuddy-io/buildbuddy/proto/firecracker"
+	repb "github.com/buildbuddy-io/buildbuddy/proto/remote_execution"
+	rnpb "github.com/buildbuddy-io/buildbuddy/proto/runner"
+	dockerclient "github.com/docker/docker/client"
+)
+
+type RunnerClients struct {
+	dockerClient   *dockerclient.Client
+	podmanProvider *podman.Provider
+}
+
+func (p *pool) initContainerProviders(env environment.Env) error {
+	if platform.DockerSocket() != "" {
+		_, err := os.Stat(platform.DockerSocket())
+		if os.IsNotExist(err) {
+			return status.FailedPreconditionErrorf("Docker socket %q not found", platform.DockerSocket())
+		}
+		dockerSocket := platform.DockerSocket()
+		if !strings.Contains(dockerSocket, "://") {
+			dockerSocket = fmt.Sprintf("unix://%s", dockerSocket)
+		}
+		p.runnerClients.dockerClient, err = dockerclient.NewClientWithOpts(
+			dockerclient.WithHost(dockerSocket),
+			dockerclient.WithAPIVersionNegotiation(),
+		)
+		if err != nil {
+			return status.FailedPreconditionErrorf("Failed to create docker client: %s", err)
+		}
+	}
```

Comment:
there is a lot of code duplicated between `runner_linux.go` and `runner_nolinux.go`, like this highlighted section.

Do you mind moving these pieces of logic to functions in `docker.go` / `podman.go` etc.

Ideally we could have a `container.Provider` interface with 2 methods `Init()` (one-time setup) / `New()` (creates a `Container` instance for a task) but for now it'd also be fine to just move this code to top-level `Init` / `New` functions in the appropriate package. (Eventually I'd like to move to the Provider interface approach and then only bind providers that are flag-enabled and compatible with the current platform. That would let us remove hacks like `firecracker_darwin.go` since the container provider setup for darwin would not depend on firecracker at all.)

Reply:
I am not happy with the current state of this PR either. 🥲
It's very MVP-ish (thus draft status).
Let me take a look at the interface `Init() / New()` approach. Sounds reasonable.

> (Eventually I'd like to move to the Provider interface approach and then only bind providers that are flag-enabled and compatible with the current platform. That would let us remove hacks like firecracker_darwin.go since the container provider setup for darwin would not need to depend on firecracker at all.)

Agree. The final state should be VFS, firecracker, podman are linux-exclusive dependencies.
And our darwin (and potentially windows) executor would not have to build with them.

Though it's gonna take a while as some of these refactoring changes are quite clunky to ship all at once.

Reply:
> Though it's gonna take a while as some of these refactoring changes are quite clunky to ship all at once.

If you decide to go with the interface approach, one way to hopefully reduce clunk factor could be to add the interface definition first + env / runner boilerplate to use the new interface, and then add the interface implementations one-by-one as separate PRs